Add markRanges method (#109)

Mottie opened the pull request:
--Updated--

#### Added `markRanges([{start: #, length: #}, {start: #, length: #}, ...], options);`

[DEMO](https://jsfiddle.net/Mottie/whkmkpfm/4/)

* Marks are applied to a range of characters derived from the text of an element.
* Pass the function an array of objects containing a start and length for each mark.
* `markRanges()` is set to automatically wrap ranges across elements. Because of this a marked range may contain more than one mark for a set range.
* Options
  * `element`
  * `className`
  * `exclude`
  * `iframes`
  * `iframesTimeout`
  * `each(node, range)` *NOTE* the `range` parameter is only available for the `markRanges()` function
  * `noMatch(range)` called if a range is invalid or ignored; returns the non-matching range object.
  * `filter(node, range, string, counter)`
  * `done()`
  * `debug` [false] - lots of logging when enabled!
  * `log`
* Throws error: if not passed an array of objects.
* Ignores & a `noMatch` callback will be executed:
  * Non-numeric values; or values that are not parsed into a number using `parseInt()`.
  * Negative values.
  * Overlapping ranges, e.g. in `[{ start:10, length:10}, {start:15, length: 15}]`, the `start: 15` range is ignored.
  * Extra elements inside of the object, e.g. in `[{start:1, length:2, foo:"bar" }]`, are ignored but passed to the `each` & `filter` callback functions.
  * Length values of zero.
  * Ranges that only capture whitespace characters. Even if the white space was wrapped, it is not visible; a debug message will show the range.
* Ignores, but still marks the range:
  * Decimal values. Values are parsed into integers; **NOTE** the `start` and `length` values passed to the `each` callback will be modified into a whole number from any strings or decimal values passed to the `markRanges` function).
  * Values larger than the target string size are set to the max string size; but the `length` value passed to the `each` & `filter` callbacks will not be modified.
* Accepts:
  * Numeric values (`10`)
  * Numbers in a string (`"10"`)
  * Non-integer numbers; these values are parsed into integers.
  * Numbers larger than the string value will end the range at the end of the context string.
* TO DO
  * Parse all DOM nodes (see #127)

julmot commented:
Thanks @Mottie. I'll come back to this this week.

Mottie commented:
This last update changed the value passed to the function into an array of objects:

```js
markRanges([{
  start: 3,
  end: 10
}, {
  start: 25,
  len: 10,
  "foo": "an extra value that is passed to the each callback function"
}], {
  each: function(node, range) {
    if (range.foo) {
      node.setAttribute("data-foo", range.foo);
    }
  }
});
```

Within the object a `start` along with either an `end` or `len` value is included. If both an `end` and `len` are included, the `end` value will be ignored.

I've updated all the tests & [demo](https://jsfiddle.net/Mottie/whkmkpfm/1/).

julmot commented on src/mark.js:
```diff
@@ -658,6 +732,88 @@ class Mark { // eslint-disable-line no-unused-vars
+     * @param {Mark~wrapRangeFromIndexEndCallback} endCb
+     * @access protected
+     */
+    wrapRangeFromIndex(ranges, filterCb, eachCb, endCb) {
```
This function is too complex/too long. I'd also vote for not adding attributes inside this function directly. All other functions are calling `wrapRangeInTextNode` and this is the only place currently. I'd vote to keep this logic and instead pass an argument that indicates that start and end attributes should be created.

Btw.: You're mixing `'` and `"` here.

Mottie replied:
> not adding attributes inside this function

Well, we can use the `each` callback to accomplish the same thing, so maybe completely removing it would be better.

julmot replied:
You mean not creating the attributes?

Mottie replied:
Yes, I can completely remove the setting of attributes because it can be done in the `each` callback... or would you prefer to keep them?

julmot replied:
I'm fine with removing it. If there's a demand for anything in that direction, we can add it later.

julmot commented on test/specs/ranges/no-match.js:
```diff
@@ -0,0 +1,89 @@
+        $ctx1 = $(".ranges-no-match > div:nth-child(1)");
+        $ctx2 = $(".ranges-no-match > div:nth-child(2)");
+        $ctx3 = $(".ranges-no-match > div:nth-child(3)");
+        // [0, 3] "should" only contain whitespace
```
Isn't this the old syntax? (comment)

Mottie replied:
On it!

Mottie commented:
I'm not sure why the tests aren't passing for you... the dependencies were updated, and I see need to be updated again. Maybe that is the problem?

As for the `manual.html` file you shared, the `main.html` looks like this:

```html
<div class="basic">⌝
1234<p>⌝
12345678Lorem ipsum dolor...
```

Counting the 2 carriage returns and 4 + 8 spaces, the "Lorem" doesn't start until the 14th position. If you change the code to this, you'll see it working:

```js
instance.markRanges([{
    start: 14,
    end: 19
}]
```

Mottie commented on src/mark.js:
```diff
@@ -61,6 +61,7 @@ class Mark { // eslint-disable-line no-unused-vars
             "caseSensitive": false,
             "ignoreJoiners": false,
             "ignoreGroups": 0,
+            "invalidMax": true,
```
LOL ok... I still haven't thought of a good one yet.

Mottie replied:
Or, should we just remove it completely and have values larger than the content automatically set to max?

julmot replied:
I think it'd be okay to remove it, if the debug log shows that is was automatically corrected. What do you think?

Mottie commented on src/mark.js:
```diff
@@ -385,6 +386,79 @@ class Mark { // eslint-disable-line no-unused-vars
+            Object.prototype.toString.call( array[0] ) !== "[object Object]"
+        ) {
+            throw new Error("markRange() will only accept an array of objects");
+        }
```
> If this check is removed...

Ok, I'll take care of it.

> I'd vote for one syntax: Either `len` or `end`.

Initially I had it set to `start` and `end` because it made sense to me, but there were two requests to include a length. Maybe I can reword it to say, use `start` and `end` or `start` and `len`. Or do you still want to completely remove one of them?

julmot replied:
What are the pros of having two separate API ways?

Mottie replied:
To me, it makes sense to use `start` and `end`.... it is closer to how a range is defined by W3C, e.g. for a [range type of input](https://www.w3.org/wiki/HTML/Elements/input/range), you set a  `min` and `max`:

```html
<input type="range" min="1" max="10">
``` 

But if third party applications return a `start` and `length`, accommodating the plugin to work with these types of values makes it easier on the developer. I know you can pass the function a `start` and `start + length` which is what I used in the [initial demo I shared](https://jsfiddle.net/Mottie/whkmkpfm/) (this demo no longer works since this branch was updated).

I don't really see any cons for including both since there is only a minimal amount of coding requires to support both.

julmot replied:
I'd vote for `length` instead of `end` due to the fact that the two that participated in the issue said that and because `substr` also needs a length: https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/String/substr

julmot replied:
I also think that renaming an object key isn't that bad if it makes our logic less complicated and the docs easier to understand.

julmot replied:
Hmm, so you're for `end`, me and two others are for `length`. We need to introduce some kind of a vote system 😆

Mottie replied:
Well, it doesn't matter to me... if we had a vote, I think the `len` would win since @DrPananoia & @olivernn asked for it in issue #109.

Mottie commented:
* Updated dependencies
* Fix comment using the older format
* Thrown error replaced with a logged message
* Completely removed the `invalidMax` option; values larger than the content length will now always stop at the end and a debug message will indicate that an adjustment was made.
* Completely removed the addition of `data-range-` attributes internally; use the `each` callback function instead. 
* Completely removed range `end` support; use `start` & `len`: e.g. `[{ start: 0, len: 5 }]`.
* Renamed range object key from `len` to `length`.

Mottie commented:
Sorry for not finishing the update yesterday... I made the fixes, but I still haven't addressed the iframe testing we discussed in Gitter... I'll spend some time on it today and then everything should be ready to go.

Mottie commented:
Done

julmot commented:
Thanks for this fantastic contribution @Mottie 🍺
